File: packages/types/src/config/storage.ts
/**
 * Storage Configuration Types
 * Unified storage configuration types for all applications
 */

/**
 * Compression Algorithm
 */
export type CompressionAlgorithm = "gzip" | "brotli";

/**
 * Compression Configuration
 */
export interface CompressionConfig {
  enabled: boolean;
  algorithm: CompressionAlgorithm;
  threshold: number;
}

/**
 * JSON Storage Configuration
 */
export interface JsonStorageConfig {
  baseDir: string;
  enableFileLock: boolean;
  compression?: CompressionConfig;
}

/**
 * SQLite Storage Configuration
 */
export interface SqliteStorageConfig {
  dbPath: string;
  enableWAL: boolean;
  enableLogging: boolean;
  readonly: boolean;
  fileMustExist: boolean;
  timeout: number;
  /** Auto-vacuum mode: NONE (default), FULL, or INCREMENTAL */
  autoVacuum?: 'NONE' | 'FULL' | 'INCREMENTAL';
  /** Journal size limit in bytes (default: 64MB, prevents unbounded WAL growth) */
  journalSizeLimit?: number;
  /** Page size in bytes (default: 4096, must be set before table creation, requires VACUUM to change) */
  pageSize?: number;
  /** Periodic maintenance interval in ms (default: 0 = disabled) */
  maintenanceIntervalMs?: number;
}

/**
 * PostgreSQL Storage Configuration
 */
export interface PostgresStorageConfig {
  /** Database host */
  host: string;
  /** Database port (default: 5432) */
  port?: number;
  /** Database user */
  username: string;
  /** Database password */
  password: string;
  /** Database name */
  database: string;
  /** Enable SSL connection (default: false) */
  ssl?: boolean;
  /** Maximum pool size (default: 20) */
  poolSize?: number;
  /** Minimum idle connections (default: 1) */
  minConnections?: number;
  /** Idle connection timeout in ms (default: 30000) */
  idleTimeout?: number;
  /** Connection timeout in ms (default: 5000) */
  connectionTimeout?: number;
  /** Maximum uses per connection before recycle (default: Infinity) */
  maxUses?: number;
}

/**
 * Storage Type
 */
export type StorageType = "json" | "sqlite" | "postgres" | "memory";

/**
 * Storage Configuration
 */
export interface StorageConfig {
  type: StorageType;
  json?: JsonStorageConfig;
  sqlite?: SqliteStorageConfig;
  postgres?: PostgresStorageConfig;
}

File: crates/wf-types/src/config/storage.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum StorageType {
    #[default]
    Sqlite,
    Postgres,
    Memory,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum CompressionAlgorithm {
    #[serde(rename = "gzip")]
    Gzip,
    #[serde(rename = "brotli")]
    Brotli,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CompressionConfig {
    pub enabled: bool,
    pub algorithm: Option<CompressionAlgorithm>,
    pub threshold: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum AutoVacuum {
    #[default]
    None,
    Full,
    Incremental,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct SqliteStorageConfig {
    pub db_path: String,
    #[serde(default)]
    pub enable_wal: bool,
    #[serde(default)]
    pub enable_logging: bool,
    #[serde(default)]
    pub readonly: bool,
    #[serde(default)]
    pub file_must_exist: bool,
    #[serde(default)]
    pub timeout: i64,
    #[serde(default)]
    pub auto_vacuum: AutoVacuum,
    pub journal_size_limit: Option<i64>,
    pub page_size: Option<i64>,
    pub maintenance_interval_ms: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PostgresStorageConfig {
    pub host: String,
    #[serde(default = "default_pg_port")]
    pub port: u16,
    pub username: String,
    pub password: String,
    pub database: String,
    #[serde(default)]
    pub ssl: bool,
    pub pool_size: Option<u32>,
    pub min_connections: Option<u32>,
    pub idle_timeout: Option<i64>,
    pub connection_timeout: Option<i64>,
    pub max_uses: Option<i64>,
}

fn default_pg_port() -> u16 {
    5432
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct StorageConfig {
    #[serde(rename = "type")]
    pub storage_type: StorageType,
    pub sqlite: Option<SqliteStorageConfig>,
    pub postgres: Option<PostgresStorageConfig>,
    /// Application name used for default db path derivation.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub app_name: Option<String>,
}

impl Default for StorageConfig {
    fn default() -> Self {
        Self {
            storage_type: StorageType::Memory,
            sqlite: None,
            postgres: None,
            app_name: None,
        }
    }
}
